src/app/clusters/operational-credentials/operational-credentials.cpp

```diff
@@ -33,33 +36,193 @@
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Failed to write admin count %" PRIX32 " in fabrics list", fabricIndex);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+AdminPairingInfo * retrieveCurrentAdmin()
+{
+    uint64_t fabricId = emberAfCurrentCommand()->source;
+    uint64_t nodeId;
+    CHIP_ERROR err = ConfigurationMgr().GetDeviceId(nodeId);
+    if (err == CHIP_NO_ERROR)
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Finding admin with fabricId  %" PRIX64 ".", fabricId);
+        return GetGlobalAdminPairingTable().FindAdmin(fabricId, nodeId);
+    }
+    return nullptr;
+}
+
+class OpCredsAdminPairingTableDelegate : public AdminPairingTableDelegate
+{
+
+    void OnAdminDeletedFromStorage(AdminId adminId) override
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Admin %" PRIX16 " was deleted from admin storage.");
+        writeAdminsIntoFabricsListAttribute();
+    }
+    void OnAdminRetrievedFromStorage(AdminId adminId, FabricId fabricId, NodeId nodeId) override
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Admin %" PRIX16 " was retrieved from storage. FabricId %" PRIX64
+                                            ", NodeId %" PRIX64 ", VendorId %" PRIX64, adminId, fabricId, nodeId);
+        writeAdminsIntoFabricsListAttribute();
+    }
+    void OnAdminPersistedToStorage(AdminId adminId, FabricId fabricId, NodeId nodeId) override
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Admin %" PRIX16 " was persisted to storage. FabricId %" PRIX64
+                                            ", NodeId %" PRIX64 ", VendorId %" PRIX64, adminId, fabricId, nodeId);
+        writeAdminsIntoFabricsListAttribute();
+    }
+};
+
+OpCredsAdminPairingTableDelegate gAdminDelegate;
+
+void emberAfPluginOperationalCredentialsServerInitCallback(void)
+{
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Initiating OpCreds cluster by writing fabrics list from admin pairing table.");
+    GetGlobalAdminPairingTable().SetAdminPairingDelegate(&gAdminDelegate);
+    writeAdminsIntoFabricsListAttribute();
+}
 
 bool emberAfOperationalCredentialsClusterRemoveFabricCallback(chip::app::Command * commandObj, chip::FabricId fabricId,
                                                               chip::NodeId nodeId, uint16_t vendorId)
 {
-    // Go through admin pairing table and remove the element
-    // Then fetch the attribute list and find the right element (or maybe just use the admin pairing table index)
-    // Then rewrite the entire attribute list (no choice for now)
     emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: RemoveFabric"); // TODO: Generate emberAfFabricClusterPrintln
-    EmberAfStatus status = EMBER_ZCL_STATUS_FAILURE;
+    
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    AdminPairingInfo * admin;
+    AdminId adminId;
+    CHIP_ERROR err;
+
+    // Fetch current admin
+    admin = retrieveCurrentAdmin();
```

Comment:
Don't we need to verify something about how this admin matches the given fabricId/nodeId/vendorId somehow?

Reply:
I do, see above

Reply:
Have a bug right now, cant seem to find the current device node id on the admin. Just using the fabricId. Not using the vendorId as it might not have been set yet.

Reply:
So the point is that we update the thing returned by retrieveCurrentAdmin to somehow reflect the "current fabric", which is why we can rely on it doing the right thing here?  That should be documented clearly; it's action-at-a-distance that is entirely non-obvious from this code.

Reply:
Sorry this was responding to retrieveCurrentAdmin. I am missing a commit to fix the removeAdmin method, as it doesn't call retrieveCurrentAdmin but find there admin matching the passed in criteria.

Reply:
fixed this